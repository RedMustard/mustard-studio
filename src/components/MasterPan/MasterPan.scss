.master-pan {
    align-items: center;
    color: #fff;
    display: flex;
    justify-content: flex-end;
    text-align: right;
}
