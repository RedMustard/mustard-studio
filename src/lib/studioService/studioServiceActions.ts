import { DispatchFunction, OscillatorId } from '../../types/types';
import { logger } from '../utils/logger/logger';

export const setMasterVolume = (value: number, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_VOLUME', payload: value });
    logger.info(`Master volume set with value ${value * 100}%`);
};

export const setMasterGainNode = (gainNode: GainNode, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_GAIN', payload: gainNode });
    logger.info('Master gain node added', gainNode);
};

export const setOscillatorVolume = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_VOLUME', payload: value });
            logger.info(`Oscillator 1 volume set with value ${value * 100}%`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_VOLUME', payload: value });
            logger.info(`Oscillator 2 volume set with value ${value * 100}%`);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator volume set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorGainNode = (gainNode: GainNode, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator gain node set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorEnabled = (isEnabled: boolean, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_ENABLED', payload: isEnabled });
            logger.info('Oscillator node 1 isEnabled: ', isEnabled);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_ENABLED', payload: isEnabled });
            logger.info('Oscillator node 1 isEnabled: ', isEnabled);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorType = (oscillatorType: OscillatorType, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_TYPE', payload: oscillatorType });
            logger.info(`Oscillator node 1 type set to ${oscillatorType}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_TYPE', payload: oscillatorType });
            logger.info(`Oscillator node 2 type set to ${oscillatorType}`);
            break;
        default:
            break;
    }
};

export const setOscillatorDetune = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_DETUNE', payload: value });
            logger.info(`Oscillator node 1 detune set to ${value}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_DETUNE', payload: value });
            logger.info(`Oscillator node 2 detune set to ${value}`);
            break;
        default:
            break;
    }
};
