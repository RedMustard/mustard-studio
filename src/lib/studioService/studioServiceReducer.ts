import { GenericAction, StudioService } from '../../types/types';

export const studioServiceReducer = (state: StudioService, action: GenericAction) => {
    switch (action.type) {
        case 'SET_MASTER_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator2: action.payload,
                },
            };

        case 'SET_MASTER_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator2: action.payload,
                },
            };

        default:
            return state;
    }
};
