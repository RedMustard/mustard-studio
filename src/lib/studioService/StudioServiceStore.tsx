import { createContext, h } from 'preact';
import { useReducer } from 'preact/hooks';
import { OSC_1_INITIAL_SETTINGS, OSC_2_INITIAL_SETTINGS, OSC_SUB_INITIAL_SETTINGS } from '../../constants';
import { DispatchFunction, StudioService } from '../../types/types';
import { deepClone } from '../utils/objects/objects';
import { studioServiceReducer } from './studioServiceReducer';


const initialState: StudioService = Object.freeze({
    master: {
        gainNode: undefined,
        panNode: undefined,
        settings: {
            volume: 0.1,
            pan: 0,
        },
    },
    oscillators: {
        osc1: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_1_INITIAL_SETTINGS,
            },
        },
        osc2: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_2_INITIAL_SETTINGS,
            },
        },
        oscSub: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_SUB_INITIAL_SETTINGS,
            },
        },
    },
});

export const getInitialState = () => deepClone(initialState);

export const StudioServiceContext = createContext<[StudioService, DispatchFunction]>([getInitialState(), () => {}]);

export const StudioServiceStore = ({ children }: any) => {
    const [state, dispatch]: [StudioService, DispatchFunction] = useReducer(studioServiceReducer, getInitialState());

    return (
        <StudioServiceContext.Provider value={[state, dispatch]}>
            {children}
        </StudioServiceContext.Provider>
    );
};
