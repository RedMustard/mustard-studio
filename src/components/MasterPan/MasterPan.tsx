import { h } from 'preact';
import { useContext, useEffect } from 'preact/hooks';
import { MAX_PAN, MIN_PAN, PAN_STEPS } from '../../constants';
import { resetMasterPanPosition, setMasterPanNode, setMasterPanPosition } from '../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../lib/studioService/StudioServiceStore';
import { ValueUnitEnum } from '../../types/runtimeTypes';
import { Fader } from '../Fader/Fader';

interface MasterPanProps {
    audioContext: AudioContext;
}

export const MasterPan = ({ audioContext }: MasterPanProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const masterPanPosition = studioService.master.settings.pan;
    const masterGainNode = studioService.master.gainNode;
    let masterPanNode = studioService.master.panNode;

    const handleOnMouseDown = (e: MouseEvent) => {
        if (e.buttons === 1 && (e.ctrlKey || e.metaKey)) {
            resetMasterPanPosition(dispatch);
        }
    };

    useEffect(() => {
        if (!masterPanNode) {
            masterPanNode = audioContext.createStereoPanner();
            setMasterPanNode(masterPanNode, dispatch);
        } else {
            masterPanNode.pan.value = masterPanPosition;
            masterPanNode.connect(audioContext.destination);
            masterGainNode.connect(masterPanNode);
        }
    }, [masterPanNode, masterPanPosition, masterGainNode]);

    return (
        <div className="master-pan">
            M Pan: &nbsp;
            <Fader
                classSuffix="master-pan"
                value={masterPanPosition}
                onInput={(value) => setMasterPanPosition(value, dispatch)}
                onMouseDown={(e: MouseEvent) => handleOnMouseDown(e)}
                maxValue={MAX_PAN}
                minValue={MIN_PAN}
                stepResolution={PAN_STEPS}
                valueUnit={ValueUnitEnum.PAN}
            />
        </div>
    );
};
