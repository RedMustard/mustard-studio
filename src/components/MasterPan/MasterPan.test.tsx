import { h } from 'preact';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { MasterPan } from './MasterPan';
import { setMasterPanPosition, setMasterPanNode, resetMasterPanPosition } from '../../lib/studioService/studioServiceActions';
import { getInitialState, StudioServiceContext } from '../../lib/studioService/StudioServiceStore';


jest.mock('../../lib/studioService/studioServiceActions');

// eslint-disable-next-line import/newline-after-import
const wamock = require('web-audio-mock-api');
const audioContext = new wamock.AudioContext();
const baseProps = {
    audioContext,
};
const mockPanNode = {
    connect: jest.fn().mockReturnThis(),
    pan: {
        value: 1.0,
    },
};
const mockGainNode = {
    connect: jest.fn().mockReturnThis(),
    gain: {
        value: 1.0,
    },
};
const initialState = {
    ...getInitialState(),
    settings: {
        master: {
            pan: 0.5,
        },
    },
    panNodes: {
        master: mockPanNode,
    },
    gainNodes: {
        master: mockGainNode,
    },
};

describe('<MasterPan />', () => {
    beforeEach(() => {
        jest.spyOn(audioContext, 'createStereoPanner').mockImplementation(() => mockPanNode);
    });
    it('renders with basic props', () => {
        const wrapper = shallow(
            <MasterPan
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('calls setMasterPanNode if masterPanNode does not exist', () => {
        shallow(
            <MasterPan
                {...baseProps}
            />,
        );
        expect(audioContext.createStereoPanner).toBeCalled();
        expect(setMasterPanNode).toBeCalled();
    });

    it('calls setMasterPanPosition when pan changed', () => {
        const wrapper = mount(
            <MasterPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('input');
        expect(setMasterPanPosition).toBeCalled();
    });

    it('sets mockPanNode.pan.value and connects to master gain node if masterPanNode exists', () => {
        mount(
            <StudioServiceContext.Provider value={[initialState, jest.fn()]}>
                <MasterPan
                    {...baseProps}
                />
            </StudioServiceContext.Provider>,
        );
        expect(mockGainNode.connect).toBeCalled();
        expect(mockPanNode.connect).toBeCalled();
        expect(mockPanNode.pan.value).toBe(initialState.settings.master.pan);
    });

    it('calls resetMasterPanPosition when ctrl clicked', () => {
        const wrapper = mount(
            <MasterPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown', { buttons: 1, ctrlKey: true });
        expect(resetMasterPanPosition).toBeCalled();
    });

    it('calls resetMasterPanPosition when cmd clicked', () => {
        const wrapper = mount(
            <MasterPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown', { buttons: 1, metaKey: true });
        expect(resetMasterPanPosition).toBeCalled();
    });

    it('does not call resetMasterPanPosition when not ctrl or cmd clicked', () => {
        const wrapper = mount(
            <MasterPan
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('mousedown', { buttons: 1 });
        expect(resetMasterPanPosition).toBeCalled();
    });
});
