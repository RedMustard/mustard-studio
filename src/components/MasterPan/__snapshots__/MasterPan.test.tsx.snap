// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MasterPan /> renders with basic props 1`] = `
<div
  className="master-pan"
>
  M Pan: Â 
  <Fader
    classSuffix="master-pan"
    maxValue={1}
    minValue={-1}
    onInput={[Function]}
    onMouseDown={[Function]}
    stepResolution={0.01}
    value={0}
    valueUnit="pan"
  />
</div>
`;
